mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -1,471 +1,504 @@
-        unsafe { self.0.as_ref().u.bdev.iovs }
+        // always free the child IO. The status of the child IO has been set by
+        // the underlying device before invocation of the callback.
+        child_io.free();
     }
 
-    /// number of iovs that are part of this IO
-    #[inline]
-    pub(crate) fn iov_count(&self) -> i32 {
-        unsafe { self.0.as_ref().u.bdev.iovcnt }
+    /// reference to the inner channels. The inner channel contains the specific
+    /// per-core data structures.
+    #[allow(clippy::mut_from_ref)]
+    fn inner_channel(&self) -> &mut NexusChannelInner {
+        NexusChannel::inner_from_channel(self.ctx().channel.as_ptr())
     }
 
-    /// offset where we do the IO on the device
-    #[inline]
-    pub(crate) fn offset(&self) -> u64 {
-        unsafe { self.0.as_ref().u.bdev.offset_blocks }
-    }
-
-    /// num of blocks this IO will read/write/unmap
-    #[inline]
-    pub(crate) fn num_blocks(&self) -> u64 {
-        unsafe { self.0.as_ref().u.bdev.num_blocks }
+    //TODO make const
+    fn data_ent_offset(&self) -> u64 {
+        let b = self.bdev_as_ref();
+        assert_eq!(b.product_name(), NEXUS_PRODUCT_ID);
+        unsafe { Nexus::from_raw((*b.as_ptr()).ctxt) }.data_ent_offset
     }
 
-    /// NVMe passthru command
-    #[inline]
-    pub(crate) fn nvme_cmd(&self) -> spdk_sys::spdk_nvme_cmd {
-        unsafe { self.0.as_ref().u.nvme_passthru.cmd }
+    /// helper routine to get a channel to read from
+    fn read_channel_at_index(&self, i: usize) -> &BdevHandle {
+        &self.inner_channel().readers[i]
     }
 
-    /// raw pointer to NVMe passthru data buffer
-    #[inline]
-    pub(crate) fn nvme_buf(&self) -> *mut c_void {
-        unsafe { self.0.as_ref().u.nvme_passthru.buf as *mut _ }
+    /// submit a read operation to one of the children of this nexus
+    #[inline(always)]
+    fn submit_read(&self, hdl: &BdevHandle) -> Result<(), Errno> {
+        let (desc, chan) = hdl.io_tuple();
+        unsafe {
+            spdk_bdev_readv_blocks(
+                desc,
+                chan,
+                self.iovs(),
+                self.iov_count(),
+                self.offset() + self.data_ent_offset(),
+                self.num_blocks(),
+                Some(Self::child_completion),
+                self.as_ptr().cast(),
+            )
+        }
+        .to_result(Errno::from_i32)
     }
 
-    /// NVMe passthru number of bytes to transfer
-    #[inline]
-    pub(crate) fn nvme_nbytes(&self) -> u64 {
-        unsafe { self.0.as_ref().u.nvme_passthru.nbytes }
+    /// submit read IO to some child
+    fn readv(&mut self) -> Result<(), Errno> {
+        if let Some(i) = self.inner_channel().child_select() {
+            let hdl = self.read_channel_at_index(i);
+            self.submit_read(hdl).map(|_| {
+                self.ctx_as_mut().in_flight += 1;
+            })
+        } else {
+            self.fail();
+            Err(Errno::ENODEV)
+        }
```

Comment:
Internal retries are done within the initiator Jan, we can rotate towards a different child if we want to in the future and I think we actually should. It is however what we had since the start so for now, to not make this any bigger, i'd suggest we keep it as is and rotate later?

Reply:
absolutely. Thanks for explaining 👍